cal-itp/data-infra
# Amplitude to Metabase feature branch

# Overall Description

This PR is where all tickets relating to https://github.com/cal-itp/data-infra/milestone/24 the `Amplitude to Metabase` milestone have been merged into, and will close https://github.com/cal-itp/data-infra/issues/960.

## What this PR includes

The following PRs were reviewed and merged into this PR:

1. https://github.com/cal-itp/data-infra/pull/1062
2. https://github.com/cal-itp/data-infra/pull/1121
3. https://github.com/cal-itp/data-infra/pull/1200
4. https://github.com/cal-itp/data-infra/pull/1209
5. https://github.com/cal-itp/data-infra/pull/1239
6. https://github.com/cal-itp/data-infra/pull/1279
- Update to the docs: https://data-infra-1116--cal-itp-previews.netlify.app/datasets_and_tables/overview.html https://data-infra-1116--cal-itp-previews.netlify.app/datasets_and_tables/amplitude.html

## To Test this PR
- `cd airflow`
- Get Amplitude API key and secret key from Amplitude production and add values to `.env`
- Run `airflow` locally
- Open http://localhost:8080/home
- Enable [amplitude_benefits](http://localhost:8080/tree?dag_id=amplitude_benefits)
- Run DAG
- Check for proper data in GCP and Metabase

## Checklist for all PRs

- [x] Run `pre-commit run --all-files` to make sure markdown/lint passes
- [x] Link this pull request to all issues that it will close using keywords (see GitHub docs about [Linking a pull request to an issue using a keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)). Also mention any issues that are partially addressed or are related.

## Airflow DAG changes checklist

- [x] Include this section whenever any change to a DAG in the `airflow/dags` folder occurs, otherwise please omit this section.
- [x] Verify that all affected DAG tasks were able to run in a local environment
- [x] Take a screenshot of the graph view of the affected DAG in the local environment showing that all affected DAG tasks completed successfully
- [x] Add/update documentation in the `docs/airflow` folder as needed
- [x] Fill out the following section describing what DAG tasks were added/updated

This PR updates the `amplitude_benefits` DAG in order to....

Adds the following DAG tasks:
- `api_to_jsonl`: Uses the `AmplitudeToFlattenedJSONOperator` to get data from Amplitude API to dataframe. Requires Amplitude API key and secret key. Saves data into the `ingest_amplitude_raw_dev` or `ingest_amplitude_raw_prod` bucket, in the `/benefits/` folder, with a `JSONL` file named with the start and end timestamps.
- `jsonl_bigquery`: Specifies data schema to ingest into BigQuery, in a table called `amplitude.benefits_events`
- `bigquery_to_views`: Users the `SqlToWarehouseOperator` to move data from BigQuery into a view, `views.amplitude_benefits_events`. Only selects the desired data fields and flattens `event_properties` and `user_properties` structs.

<img width="1504" alt="image" src="https://user-images.githubusercontent.com/3673236/160724091-7a7be544-6c24-4253-b71f-206bc5942439.png">

<img width="1501" alt="image" src="https://user-images.githubusercontent.com/3673236/160724166-0cc628ba-eaa4-41b8-b30a-30a54e90fc1a.png">
_Last DAG runs are all green_


Updates the following:

- Add `is_development()` to macros, available to use in operators
- Adds `airflow/.env.sample`
- Adds 2 variables to `airflow/docker-compose.yaml`
